src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php

```diff
@@ -144,4 +144,20 @@ private function processForm(FormHandlerInterface $formHandler, Request $request
 
         return true;
     }
+
+    /**
+     * Generates PDF of given invoice ID.
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))", message="Access denied.")
+     *
+     * @param int $invoiceId
+     */
+    public function generatePdfByIdAction(int $invoiceId)
+    {
+        $this->get('prestashop.adapter.pdf.generator.single_invoice')->generatePDF([$invoiceId]);
+
+        // When using legacy generator,
+        // we want to be sure that displaying PDF is the last thing this controller will do
+        die();
```

Comment:
I'm afraid of side effect. Could you use BinaryResponse? or if needed can be catch by an ob_start() :thinking:

Reply:
https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php#L284-L291 I made this same way like orders controller do.

Reply:
What the fuuu*** :scream: